### ISSUE TITLE
Consider a registry of native peers

### ISSUE THREAD

**Issue by dmitry-timofeev** [prevoty/jni-rs#84]
## Overview
This is a _feature request_ for an additional safety net for the use-case when Java passes pointers to native objects to the native code.

Consider the following code:
```java
// An example of Java proxy of a native object: `vec` proxy
class VecProxy {
    // A handle to the corresponding native object (aka native peer).
    private final long nativeHandle;

    // Constructors, public methods using the native methods below:
    
    // Creates a new `vec` and returns a pointer to it.
    private static native long nativeCreate(int initialSize);

    // Adds the given element to the vector.
    private static native void nativeAdd(long nativeHandle, int newElement);

    // Destroys the corresponding native peer.
    private static native void nativeDestroy(long nativeHandle);
}
```

The native code has to cast `jlong`s to the corresponding objects. If there is a bug in Java code (use-after-free, double-free, unknown-pointer, wrong-pointer-type), the app is likely to crash.

## Possible solution
A registry of all native objects that can be used at the given moment by a pointer. Such an object checks that a native handle:
  - is known,
  - has a correct type.

Such checks may be optional (e.g., enabled in debug builds only), and may be exposed through such interface:
  - Create a handle to a native object, optionally, registering it in the registry: `to_handle`
  - Cast a handle to a native object of the given type (with _optional_ checks above): `cast_handle`
  - Drop handle, destroying the native object and, optionally, removing the handle from the registry: `drop_handle`

In our code base, we currently have a synchronized HashMap-based implementation.

## Questions
- [ ] Q0: Is it useful to all users? Shall it be:
    - In the core library?
    - In a separate satellite crate, e.g., `jni-java-proxies`?
    - Outside the lib?

**Comment by dmitry-timofeev** [comment]
> Did you consider, as a possible solution, the usage of generation-based (also known as versioned) pointers? This seems like a possible (and possibly efficient) solution to the issue.

No, have not had a change to use them. Are they supposed to be significantly more efficient than, say, a proper concurrent hash map, allowing both concurrent reads *and* writes?

Our [present code](https://github.com/exonum/exonum-java-binding/blob/v0.3/exonum-java-binding-core/rust/src/utils/resource_manager/imp.rs#L22) uses a simple mutex-protected HashMap, but we consider using a concurrent hash map. However, this feature is mostly oriented at developers working on the project to catch bugs during development early, so performance is not an issue.

**Comment by koutheir** [comment]
> Are they supposed to be significantly more efficient than, say, a proper concurrent hash map, allowing both concurrent reads and writes?

I would say they can be more efficient, but benchmarking is the ultimate answer.

**Comment by dmitry-timofeev** [comment]
> Q0: Is it useful to all users?

Given our experience in EJB; some feedback from people who got this wrong; and the feedback from our tutorial project, I think yes, it shall be added, either as a sister crate, or as a module within this crate.

A battle-tested implementation already exists in EJB under Apache 2.0, so we can pull it from here, with an appropriate notice on 3rd-party code.

**Comment by cherryleafroad** [comment]
I'd appreciate something like this honestly. It definitely would make things safer / easier when dealing with native pointers
